client/state/plugins/installed/selectors.ts

```diff
@@ -0,0 +1,382 @@
+}
+
+export function isRequesting( state: AppState, siteId: number ) {
+	if ( typeof state.plugins.installed.isRequesting[ siteId ] === 'undefined' ) {
+		return false;
+	}
+	return state.plugins.installed.isRequesting[ siteId ];
+}
+
+export function isRequestingForSites( state: AppState, sites: number[] ) {
+	// As long as any sites have isRequesting true, we consider this group requesting
+	return some( sites, ( siteId ) => isRequesting( state, siteId ) );
+}
+
+export function isRequestingForAllSites( state: AppState ) {
+	return state.plugins.installed.isRequestingAll;
+}
+
+export function requestPluginsError( state: AppState ) {
+	return state.plugins.installed.requestError;
+}
+
+const getSiteIdsThatHavePlugins = createSelector(
+	( state: AppState ) => {
+		return Object.keys( state.plugins.installed.plugins ).map( ( siteId ) => Number( siteId ) );
+	},
+	( state ) => [ state.plugins.installed.plugins ]
+);
+
+/**
+ * The server returns plugins store at state.plugins.installed.plugins are indexed by site, which means
+ * that the information for a plugin may be spread across multiple site objects. This selector transforms
+ * that structure into one indexed by the plugin slugs and memoizes that structure.
+ */
+export const getAllPluginsIndexedByPluginSlug = createSelector(
+	( state: AppState ) => {
+		if ( isRequestingForAllSites( state ) ) {
+			return {};
+		}
+
+		return getSiteIdsThatHavePlugins( state ).reduce(
+			( plugins: { [ key: string ]: Plugin }, siteId ) => {
+				if ( isRequesting( state, siteId ) ) {
+					return plugins;
+				}
+
+				const installedPlugins = state.plugins.installed.plugins as InstalledPlugins;
+
+				const pluginsForSite = installedPlugins[ siteId ] || [];
+				pluginsForSite.forEach( ( plugin: InstalledPluginData ) => {
+					const sitePluginProperties = [ 'active', 'autoupdate', 'update', 'version' ];
+					const sitePluginInfo = pick( plugin, sitePluginProperties );
+					const otherPluginInfo = omit( plugin, sitePluginProperties );
+
+					plugins[ plugin.slug ] = {
+						...plugins[ plugin.slug ],
+						...otherPluginInfo,
+						sites: {
+							...plugins[ plugin.slug ]?.sites,
+							[ siteId ]: sitePluginInfo,
+						},
+					};
+				} );
+				return plugins;
+			},
+			{}
+		);
+	},
+	( state ) => [
+		isRequestingForAllSites( state ),
+		getSiteIdsThatHavePlugins( state ),
+		...getSiteIdsThatHavePlugins( state ).map( ( siteId ) => isRequesting( state, siteId ) ),
+	]
+);
+
+/**
+ * The plugins here differ from the plugin objects found on state.plugins.installed.plugins in that each plugin
+ * object has data from all the different sites on it, whereas on state.plugins.installed.plugins they only have
+ * the state for the site which index they are under. The objects here are the same as those returned from
+ * getAllPluginsIndexedByPluginSlug, except they are indexed by siteId.
+ */
+export const getAllPluginsIndexedBySiteId = createSelector(
```

Comment:
Originally it was just performance improvements for the selectors. `getFilteredAndSortedPlugins` was `getPlugins` and it was very slow algorithmically.

I created `getAllPluginsIndexedBySiteId` and `getAllPluginsIndexedByPluginSlug` as internal helpers to provide some data structures that would reduce the algorithmic complexity in `getPlugins`. Then I noticed that `getPlugins` was being used in many places to find a single plugin by searching the array of all plugins which was part of the performance issues, and my helper selectors were memoized and find a single plugin in constant time so I exposed them just for that use case.

I then spent a lot of time tracking down minor bugs which were very difficult to find because they were happening in code far away from these selectors, usually in a component that had data from the selector passed down through many components and modified by some until finally the bug manifested. In almost every case it was a type error because some javascript code either changed the type of an object or used a type that didn't exist. This took enough time and I found enough of these bugs that I decided to change it to TypeScript which did fix them all at once and will prevent them from happening again.

At this point I don't think that separating the two is possible, or at least not worth the effort. I would have to take the existing TypeScript code and convert it back into JavaScript in order to keep the bug fixes, and then convert it back to TypeScript in another PR, and it just doesn't seem worth it.

Reply:
I understand your rationale, but I'd disagree with the approach taken; I think the approach here to do all refactors in the same PR has created a potentially incomprehensible change that we just should not merge altogether, but rather spend some effort splitting changes out into meaningful, atomic changes. What you described seemed to me like at least a few PRs we could have undertaken one by one. I realize this takes more time, but it's just the more responsible way to work with a huge codebase spanning hundreds of developers, hundreds of projects, almost ten years of work, and thousands of commits. My only advice here is to be responsible and push changes in a way that our future selves will thank our past selves for. I'm happy to provide ideas on how to split the changes if you're struggling. I did provide a rough plan in the primary comment of my previous review, but I'm happy to take it to the lower level if that will help.